### PR TITLE
listen on chain event and appEvent$ for token balance update

### PR DESCRIPTION
**Pull request by ShawTim**

### CODE REVIEW

File: packages/tempus-core-services/src/services/ERC20TokenService.ts [hunk 1]
```diff
@@ -154,7 +154,8 @@ export class ERC20TokenService {
     if (this.contract) {
       // In case of ETH trigger transfer event on every new block
       if (this.contract.address === ZERO_ETH_ADDRESS) {
-        this.contract.provider.on('block', listener);
+        // skip this. native token is not a ERC20 contract and cannot listen for transfer events
+        // this.contract.provider.on('block', listener);
```
Comment (danieledamiani) [inline packages/tempus-core-services/src/services/ERC20TokenService.ts:158]: we could refactor this block like 
```
if (this.contract.address !== ZERO_ETH_ADDRESS) {
...
```
and add the comment just above

File: packages/tempus-client_v3/src/hooks/useTokenBalance.ts [hunk 2]
```diff
@@ -71,27 +245,28 @@ const stream$ = combineLatest([walletAddress$, selectedChain$, servicesLoaded$])
...
-
-      return balanceFetch$.pipe(
-        map(tokenBalance => ({
-          [`${tokenBalanceData.chain}-${tokenBalanceData.address}`]: {
-            balance: tokenBalance,
-            address: tokenBalanceData.address,
-            chain: tokenBalanceData.chain,
-          },
-        })),
+      const balanceBalance$ = fetchData(tokenBalanceData.chain, tokenBalanceData.address, walletAddress);
```
Comment (danieledamiani) [inline packages/tempus-client_v3/src/hooks/useTokenBalance.ts:248]: `balanceBalance`?

File: packages/tempus-client_v3/src/hooks/useTokenBalance.ts [hunk 3]
```diff
@@ -71,27 +245,28 @@ const stream$ = combineLatest([walletAddress$, selectedChain$, servicesLoaded$])
...
+      return balanceBalance$.pipe(
+        map(tokenBalance =>
+          tokenBalance
```
Comment (danieledamiani) [inline packages/tempus-client_v3/src/hooks/useTokenBalance.ts:251]: this should be done by a `filter` operator

File: packages/tempus-client_v3/src/hooks/useTokenBalance.ts [hunk 4]
```diff
@@ -51,13 +60,178 @@ tokenList.forEach(token => {
   });
 });
 
+const fetchData = (chain: Chain, tokenAddress: string, walletAddress: string): Observable<Decimal | null> => {
```
Comment (danieledamiani) [inline packages/tempus-client_v3/src/hooks/useTokenBalance.ts:63]: this should be named something like `userTokenBalance$`

File: packages/tempus-client_v3/src/hooks/useTokenBalance.ts [hunk 5]
```diff
@@ -51,13 +60,178 @@ tokenList.forEach(token => {
...
+    console.error(
+      `useTokenBalances - cannot fetch token balance for ${tokenAddress} on ${chain} for $wallet ${walletAddress}`,
+      error,
+    );
+    return of(null);
+  }
+};
+
+// stream for listening chain events
+const txnStream$ = combineLatest([walletAddress$, signer$, servicesLoaded$]).pipe(
```
Comment (danieledamiani) [inline packages/tempus-client_v3/src/hooks/useTokenBalance.ts:85]: `transactions$`?

File: packages/tempus-client_v3/src/hooks/useTokenBalance.ts (hunk 4 above)
Comment (ShawTim) [reply packages/tempus-client_v3/src/hooks/useTokenBalance.ts:63]: but this is a function to fetch the user balance and to parse the response as `Observable`. it's not an `Observable` itself

File: packages/tempus-client_v3/src/hooks/useTokenBalance.ts (hunk 3 above)
Comment (ShawTim) [reply packages/tempus-client_v3/src/hooks/useTokenBalance.ts]: updated and applied on others

File: packages/tempus-client_v3/src/hooks/useTokenBalance.ts [hunk 6]
```diff
@@ -51,13 +60,178 @@ tokenList.forEach(token => {
...
+      yieldBearingTokenBalanceMap$,
+      yieldsTokenBalanceMap$,
+      ammTokenBalanceMap$,
+    );
+  }),
+);
+
 // Stream that goes over all tokens and fetches their balance, this happens only when wallet address changes
-const stream$ = combineLatest([walletAddress$, selectedChain$, servicesLoaded$]).pipe(
+const initStream$ = combineLatest([walletAddress$, selectedChain$, servicesLoaded$]).pipe(
```
[the remark this answers is not in the document]
Comment (ShawTim) [reply packages/tempus-client_v3/src/hooks/useTokenBalance.ts]: `walletStream$`

File: packages/tempus-client_v3/src/hooks/useTokenBalance.ts (hunk 4 above)
Comment (ShawTim) [reply packages/tempus-client_v3/src/hooks/useTokenBalance.ts:63]: this should be consistent with other hooks, which having a `fetchData` function that returns an `Observable` results

File: packages/tempus-client_v3/src/hooks/useTokenBalance.ts (hunk 5 above)
Comment (ShawTim) [reply packages/tempus-client_v3/src/hooks/useTokenBalance.ts]: `transactionStream$`

File: packages/tempus-client_v3/src/hooks/useTokenBalance.ts [hunk 7]
```diff
@@ -51,13 +60,163 @@ tokenList.forEach(token => {
...
+            }),
+          );
+        };
+
+        // TODO: should we call offTransfer()? this stream will only run when wallet changes anyway
+        erc20TokenService.onTransfer(walletAddress, null, transferListener);
+        erc20TokenService.onTransfer(null, walletAddress, transferListener);
+
+        return subject$.pipe(
+          filter(tokenBalance => Boolean(tokenBalance)),
```
Comment (lazarmitic) [inline packages/tempus-client_v3/src/hooks/useTokenBalance.ts:112]: we should update balance here event if it's already present.

File: packages/tempus-client_v3/src/hooks/useTokenBalance.ts [hunk 8]
```diff
@@ -33,6 +38,10 @@ interface TokenBalanceData {
   chain: Chain;
 }
 
+interface TokenBalanceMap {
+  [chainPoolAddress: string]: TokenBalance;
```
[the remark this answers is not in the document]
Comment (ShawTim) [reply packages/tempus-client_v3/src/hooks/useTokenBalance.ts]: fixed

**Comment by iamvukasin** [comment]
Reviewing

File: packages/tempus-client_v3/src/hooks/useTransactions.ts [hunk 9]
```diff
@@ -0,0 +1,55 @@
+import { JsonRpcSigner } from '@ethersproject/providers';
+import { BehaviorSubject, combineLatest, filter, merge, mergeMap, Observable, of } from 'rxjs';
+import { getDefinedServices } from 'tempus-core-services';
+import { servicesLoaded$ } from './useServicesLoaded';
+import { tokenList$ } from './useTokenList';
+import { signer$ } from './useSigner';
+import { walletAddress$ } from './useWalletAddress';
+import { TokenListItem } from '../config/ConfigManager';
+
+export interface TrasnferTransaction {
```
Comment (iamvukasin) [inline packages/tempus-client_v3/src/hooks/useTransactions.ts:10]: Typo
```suggestion
export interface TransferTransaction {
```

Comment (ShawTim) [reply packages/tempus-client_v3/src/hooks/useTransactions.ts]: fixed

**Review by iamvukasin — APPROVED** [review]

**Pull request merged**